schemas/database/013_batches_and_transactions.sql

```diff
@@ -0,0 +1,24 @@
+-- Create ENUM for prover type
+CREATE TABLE batches
+(
+    id         BIGSERIAL UNIQUE PRIMARY KEY,
+    next_root  BYTEA       NOT NULL UNIQUE,
+    prev_root  BYTEA UNIQUE,
+    created_at TIMESTAMPTZ NOT NULL,
+    batch_type VARCHAR(50) NOT NULL,
+    data       JSON        NOT NULL,
+
+    FOREIGN KEY (prev_root) REFERENCES batches (next_root) ON DELETE CASCADE
```

Comment:
My thought here was at first to ensure we have consistent data in DB. As you mentioned without option to check values other than the ones in a row being inserted it is hard. And this constraint is quite important in terms of HA with multiple instances running same time trying to insert data to DB. Now only one will win and second query will be rejected. With this implementation that uses foreign key (linked list) we can also be sure that `id` here allows to order the table properly.  It is because you can't insert batch without having previous one already in table. And that means the id sequence value was incremented already and newly generated one for that statement must be greater. We are using it here for example: https://github.com/worldcoin/signup-sequencer/pull/731/files#diff-23b4c110a535fa6df354073bab85cf1645400d08a9a202311db1d2898dc406bbR1010

Reply:
Yep makes total sense. My only concern it that traversal becomes really difficult if we even need to do that. Perhaps something like this would be more appropriate:
```sql
CREATE OR REPLACE FUNCTION validate_prev_root_consecutive()
RETURNS TRIGGER AS $$
DECLARE
    previous_next_root BYTEA;
    previous_id BIGINT;
BEGIN
        -- Get the previous row's next_root and id
        SELECT next_root, id INTO previous_next_root, previous_id
        FROM batches
        WHERE id = NEW.id - 1;
       -- Might need some null/0 checks

        -- Check if the prev_root matches the previous row's next_root
        IF NEW.prev_root IS DISTINCT FROM previous_next_root THEN
            RAISE EXCEPTION 'prev_root does not match the previous row''s next_root';
        END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER validate_prev_root_consecutive_trigger
BEFORE INSERT ON batches
FOR EACH ROW
EXECUTE FUNCTION validate_prev_root_consecutive();
```

Reply:
Could elaborate more on what is the change? With this trigger I see two problems:
* checking if id = prev_id + 1 -> sometimes sequence is increased when accessed for inserts and then rejected. I guess we will then manually put next id calculated based on previous id.
* It is created only for insert, what about updates or some other scenarios which may occur in the future? Deletes for example?
And just in general: how it makes traversal easier? I mean now we can sort on id with current solution and have things in proper order for traversing. What am I missing with this change? ;)

Reply:
@piohei Yeah the more I think about this I think I'm realizing why you've done it like this. I hadn't really considered deletions. I think no matter what you do, in order to traverse the list in order you need a recursive CTE.

I think it would still be a minor improvement to make to change prev_root to parent_id, place the foreign key on that, and validate with a trigger. You'd save some storage that way, but it won't help with traversal the way I thought it might. So I'm not sure that change is actually worth it :p